lib/DBus/Internal/Wire.hs

```diff
@@ -346,17 +350,25 @@ unmarshalDouble = unmarshalGet 8
     getFloat64le
 
 marshalUnixFd :: Fd -> Marshal ()
-marshalUnixFd (Fd x)
+marshalUnixFd fd@(Fd x)
     | x < 0 = throwError ("Invalid file descriptor: " ++ show x)
     | toInteger x > toInteger (maxBound :: Word32) = throwError ("D-Bus forbids file descriptors exceeding UINT32_MAX: " ++ show x)
-    | otherwise = marshalWord32 (fromIntegral x)
+    | otherwise = do
+        MarshalState builder count fds <- getState
+        case Data.Map.lookup fd fds of
+            Just i -> marshalWord32 i
+            Nothing -> do
+                let i = fromIntegral (Data.Map.size fds)
+                putState (MarshalState builder count (Data.Map.insert fd i fds))
+                marshalWord32 i
 
 unmarshalUnixFd :: Unmarshal Fd
 unmarshalUnixFd = do
-    x <- unmarshalWord32
-    when (toInteger x > toInteger (maxBound :: CInt))
-        (throwError ("Invalid file descriptor: " ++ show x))
-    return (Fd (fromIntegral x))
+    x <- fromIntegral <$> unmarshalWord32
+    UnmarshalState _ _ fds <- getState
+    when (x >= length fds) $ do
+        throwError ("File descriptor index " ++ show x ++ " out of bounds - only " ++ show (length fds) ++ " file descriptors in message header.")
+    return (fds !! x)
```

Comment:
`fds` is likely to be pretty small (each read call can return no more than 16: see `recvWithFds`), so although `(!!)` is O(n) it does not really matter.